### ISSUE TITLE
memory raise rapidly while my dataSourceName is invalid

### ISSUE THREAD

**Issue by wanfuuyu** [rana/ora#207]
memory raise rapidly while my dataSourceName is invalid,but no problems when my dataSourceName is correct 

I use the same code pattern in my mysql project using "github.com/go-sql-driver/mysql",and no such problem

following is my testing code

```go

package main

import (
	"database/sql"
	"net/http"
	_ "net/http/pprof"

	"github.com/prometheus/common/log"
	_ "gopkg.in/rana/ora.v4"
)

func main() {
	connStr := `system/oracle@(DESCRIPTION=(ADDRESS_LIST=(ADDRESS=(PROTOCOL=TCP)(HOST=10.211.55.2)(PORT=1521)))(CONNECT_DATA=(sid=xe)))`
	db, _ := sql.Open("ora", connStr)
	defer db.Close()
	http.HandleFunc("/", func(writer http.ResponseWriter, request *http.Request) {
		if err := db.Ping(); err != nil {
			log.Errorln(err)
		}
	})
	http.ListenAndServe(":9103", nil)
}

```

**Comment by tgulacsi** [comment]
Just check your errors, as everywhere, everytime - or take the consequences.

**Comment by wanfuuyu** [comment]
But how can I prevent it just through check errors

**Comment by tgulacsi** [comment]
In your example you skip checking  the error returned by `sql.Open`. 
You have to check and handle it:
```
import "github.com/pkg/errors"

...

db, err := sql.Open("ora", connStr)
if err != nil {
    log.Fatal(errors.Wrap(err, connStr))
}
```

**Comment by wanfuuyu** [comment]
**Thank you for your patience**

but just like i said, nil error returns in `sql.Open`, I saw it in my previous sample which is different from 
the list one

Do you mean even if the `sql.Open` returns non-nil error, as long as my dataSourceName is incorrect ,everytime I try to Ping it will open a new connection, but Pings will share one connection when the dataSourceName is correct

**Comment by tgulacsi** [comment]
Yes. database/sql.Open returns a connection pool, which reuses connections as possible, but opens new connections when required - for example if there's no successful connections, it will open a new one - for each ping, or each use in case of bad connStr.

**Comment by wanfuuyu** [comment]
😩  crying...
Is there any way to release the memory caused by connections with bad connStr except restarting my main process  ? ?

**Comment by tgulacsi** [comment]
Have you tried db.Close ?
I mean, Ping the new connStr, and Close on error.

**Comment by kindule** [comment]
Hi, @tgulacsi the db close is already used. and we test the code with mysql driver and no memory leak occur

**Comment by tgulacsi** [comment]
Can you tweak TestOpenBadMemoryIssue207 in z_environment_test.go at e52c26d to reproduce the issue?

**Comment by wanfuuyu** [comment]
# **Thanks**
###  **Here is my running test based on z_environment_text.go**

--- FAIL: TestOpenBadMemoryIssue207 (0.04s)
       z_environment_test.go:21: Allocated 0: 121856
	z_environment_test.go:32: Allocated 1: 122448
	z_environment_test.go:32: Allocated 2: 123264
	z_environment_test.go:32: Allocated 3: 124256
	z_environment_test.go:32: Allocated 4: 125024
	z_environment_test.go:32: Allocated 5: 125632
       ...
	z_environment_test.go:32: Allocated 95: 159872
	z_environment_test.go:32: Allocated 96: 159872
	z_environment_test.go:32: Allocated 97: 160224
	z_environment_test.go:32: Allocated 98: 159520
	z_environment_test.go:32: Allocated 99: 159872
	z_environment_test.go:32: Allocated 100: 159872
	z_environment_test.go:35: atlast: 38016
	z_environment_test.go:37: Consumed more than 32KiB of memory: 38016
FAIL
FAIL	command-line-arguments	0.053s

### **But I till don't know the relationship between this test and my issue, what I care is that memory stay high even if I tried `db.Close` after  pinging the bad connStr**

**Comment by tgulacsi** [comment]
I don't know your issue in detail, that's why I presented a test (skeleton), and asked you to fill in whatever you want, to reproduce your problem.
Open & Close is not enough - I've added a Ping, but that locked me out, so I had to modify the test to alter the user name, too.
That's because you haven't detailed what kind of connection string you're using which is "bad".

Please see 95d4c1f and try again!

The questions we need to answer:
1. can your memory problem be inspected with runtime.ReadMemStats, or just by "top" ?
2. Is Open, Ping, Close enough to enforce the error?
3. how much memory is leaking by each cycle?

**Comment by wanfuuyu** [comment]
Here is the final str to open, I picked a wrong sid

`system/oracle@(DESCRIPTION=(ADDRESS_LIST=(ADDRESS=(PROTOCOL=TCP)(HOST=10.211.55.12)(PORT=1521)))(CONNECT_DATA=(sid=x)))`

following is the testing log, It seems nothing weird 

    --- FAIL: TestOpenBadMemoryIssue207 (0.25s)
	z_environment_test.go:21: Allocated 0: 122272
	z_environment_test.go:30: Env.OpenCon Env.OpenSrv Env.ociError ORA-12505: TNS:listener does not currently know of SID given in connect descriptor
		
	z_environment_test.go:35: Allocated 1: 123456
	z_environment_test.go:30: Env.OpenCon Env.OpenSrv Env.ociError ORA-12505: TNS:listener does not currently know of SID given in connect descriptor
		
	z_environment_test.go:35: Allocated 2: 124512
	z_environment_test.go:30: Env.OpenCon Env.OpenSrv Env.ociError ORA-12505: TNS:listener does not currently know of SID given in connect descriptor
		
	z_environment_test.go:35: Allocated 3: 125536
	z_environment_test.go:30: Env.OpenCon Env.OpenSrv Env.ociError ORA-12505: TNS:listener does not currently know of SID given in connect descriptor
		
       . . . .        

	z_environment_test.go:35: Allocated 98: 180960
	z_environment_test.go:30: Env.OpenCon Env.OpenSrv Env.ociError ORA-12505: TNS:listener does not currently know of SID given in connect descriptor
		
	z_environment_test.go:35: Allocated 99: 181024
	z_environment_test.go:30: Env.OpenCon Env.OpenSrv Env.ociError ORA-12505: TNS:listener does not currently know of SID given in connect descriptor
		
	z_environment_test.go:35: Allocated 100: 181792
	z_environment_test.go:38: atlast: 59520
	z_environment_test.go:40: Consumed more than 32KiB of memory: 59520

    FAIL
    FAIL	command-line-arguments	0.270s


Here is the thing, when I use top command to watch the res of this process, it raised 0.1-0.3m each Open-Ping-Close circle.

**Comment by tgulacsi** [comment]
Can you try & post the results of 

    go test -c && /usr/bin/time ora.v4.test -run=Issue207 -test.v

both with 95d4c1f and a3970f1 ?
In the latter I've tried to free all allocated OCI handles on error in OpenSrv (env.go)

**Comment by wanfuuyu** [comment]
Hi dude !!!! 

I both use 95d4c1f and a3970f1.  95d4c1f won't help, res stay going high. But when I use  a3970f1 , I  found that res memory just raise a little at the begain of my bad connStr test, and stay stable after that, thinking my issue is resolved in  a3970f1.

I will add more test case to ensure there is no more hidden trouble, once I found it, I will come here again..

Thanks..

**Comment by tgulacsi** [comment]
I've released it as v4.1.10.
